go-controller/pkg/ovn/base_network_controller_pods.go

```diff
@@ -985,41 +992,126 @@ func (bnc *BaseNetworkController) shouldReleaseDeletedPod(expectedSwitchName, sw
+	// release the IP
+	if bnc.allocatesPodAnnotation() {
 		return true, nil
 	}
 
-	// check to make sure no other pods are using this IP before we try to release it if this is a completed pod.
-	shouldRelease, err := bnc.lsManager.ConditionalIPRelease(switchName, podIfAddrs, func() (bool, error) {
+	// if this pod applies to live migration, it could have migrated do not filter node name
+	nodeName := ""
+	if !kubevirt.IsPodLiveMigratable(pod) {
+		nodeName = pod.Spec.NodeName
+	}
 
-		// Ignore pods on other switches
-		if expectedSwitchName != switchName {
-			return true, nil
-		}
+	shouldRelease, err := bnc.canReleasePodIPs(podIfAddrs, nodeName)
+	if err != nil {
+		return false, err
+	}
 
-		// if this pod applies to live migration, it could have migrated do not filter node name
-		nodeName := ""
-		if !kubevirt.IsPodLiveMigratable(pod) {
-			nodeName = pod.Spec.NodeName
+	return shouldRelease, nil
+}
+
+// trackReleasedPodsBeforeStartup tracks from all of the annotated pods on startup,
+// which ones might have had their IPs previously released. This information is
+// used to prevent those IPs to be released while they might be still used by
+// other running pods and is tracked per NAD, as it is possible for a pod to
+// have been released for one of its NADs but not all in an unexpected error
+// condition. It uses the following rules:
+//   - One or more completed pods sharing an IP with a running pod are
+//     considered released.
+//   - One or more completed pods sharing an IP are considered released except
+//     the last one to be initialized.
+func (bnc *BaseNetworkController) trackReleasedPodsBeforeStartup(podAnnotations map[*kapi.Pod]map[string]*util.PodAnnotation) {
+	// we will order the pods by order of initialization, by that time pods
+	// should have been already allocated exclusive IPs
+	getInitializedConditionTime := func(pod *kapi.Pod) time.Time {
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == kapi.PodInitialized {
+				return condition.LastTransitionTime.Time
+			}
 		}
-		canRelease, err := bnc.canReleasePodIPs(podIfAddrs, nodeName)
-		if err != nil {
-			return false, fmt.Errorf("unable to determine if completed pod IP is in use by another pod. "+
-				"Will not release pod %s/%s IP: %#v from allocator. %v", pod.Namespace, pod.Name, podIfAddrs, err)
+		return time.Time{}
+	}
+
+	// order the pods, running pods first, then completed pods ordered by
+	// initialization time
+	pods := maps.Keys(podAnnotations)
+	sort.Slice(pods, func(i, j int) bool {
+		if !util.PodCompleted(pods[i]) {
+			return true
 		}
+		return getInitializedConditionTime(pods[i]).After(getInitializedConditionTime(pods[j]))
+	})
 
-		if !canRelease {
-			klog.Infof("Will not release IP address: %s for %s. Detected another pod using it.",
-				util.JoinIPNetIPs(podIfAddrs, " "), podDesc)
-			return false, nil
+	bnc.releasedPodsBeforeStartup = map[string]sets.Set[string]{}
```

Comment:
I didn't do it because this is running from `syncPods` and we only have one thread at that time. But if you think we should for consistency I will add that.